Added extra tests for most features

schneems opened the pull request:
Added extra tests for most features

Most features are now tested.

schneems committed:
readme

schneems committed:
Added extra tests for most features

Most features are now tested.

schneems committed:
Added extra tests for most features

Most features are now tested.

msakrejda commented on README.md:
```diff
@@ -28,15 +28,34 @@ Create a `PgLock.new` instance and call the `lock` method to ensure exclusive ex
+
+## Session based locking
+
+The postgres lock is unique across different database sessions, if the same session tries to aquire the same lock it will succeed. So while `PgLock` will guarantee unique execution across machines and processes, it will not block the same process (sharing the same connection session) from running. For example while you would think the middle block would not run in this example:
```
Maybe mention the term`reentrant` here somewhere, since that's [the technical term](https://en.wikipedia.org/wiki/Reentrant_mutex) people may be expecting?

msakrejda commented on README.md:
```diff
@@ -28,15 +28,34 @@ Create a `PgLock.new` instance and call the `lock` method to ensure exclusive ex
+key = "all_your_base"
+PgLock.new(name: key).lock do
+  puts "First block called"
+  PgLock.new(name: key).lock do
```
Is this guaranteed to give you the same underlying session? If not, this may actually deadlock, since the second session (for the inner `PgLock`) may wait for the same advisory lock that the session of the outer `PgLock` just acquired.

That is, in #2, I talk about the problem from the perspective of different threads possibly sharing the same underlying session due to the connection pool, but there's also the reverse problem: the same thread/process can be given a _different_ connection by the pool (e.g., if the pool uses a round-robin mechanism to hand out connections, and the pool includes more than one connection).

schneems replied:
Active record provides a single connection for the duration of a request https://github.com/rails/rails/blob/76698441453821fb175799b99099322688bde6c1/activerecord/lib/active_record/query_cache.rb#L31 (if you're using the "query cache" middleware which is default in Rails). If you're on the same connection it's my understanding that you're in the same session, though maybe there are edge cases where you could manually disconnect from the session manually.

schneems committed:
default rake task

schneems committed:
more helpful error output for tests

schneems committed:
Redirect stdout and err to file

schneems committed:
More error messages

schneems committed:
Fix tests

Moved the lock! code to running in an Open3 process instead of a Process.spawn, because...i don't know but it fixed the test.

Improved the method of counting occurrences of strings in various logs and other outputs.
